Add the SGSR 1.0 upsampling algorithm

HH45137 opened the pull request:
Add Snapdragon Game Super Resolution 1.0 upsampling algorithm

[Snapdragon™ Game Super Resolution](https://github.com/SnapdragonStudios/snapdragon-gsr)

HH45137 committed:
Add the SGSR 1.0 upsampling algorithm

Add SGSR 1.0 upsampling algorithm

Snapdragon Game Super Resolution 1.0 upsampling algorithm

Enable the UseEdgeDirection macro of SGSR 1.0

If set, will use edge direction to improve visual quality expect a minimal cost increase

Add the SCALING_3D_MODE_SGSR1 macro to the script

Update the edge sharpness value of SGSR1

Fix the sorting bug of VIEWPORT_SCALING_3D_MODE_SGSR1_SPATIAL

HH45137 committed:
Fix the issue of GLSL code format checking

HH45137 commented:
> Does it work on android phones? Would be cool to have a mobile alternative to fsr.
Thank you. Currently, only Forward+ is supported.

HH45137 committed:
Fix ensure compatibility

Fix the adjust the position of the macro to ensure compatibility

Add the SGSR1 entry to the document

Fix the sequence issue of the SGSR1 option

Bromeon commented on scene/main/viewport.h:
```diff
@@ -99,6 +99,7 @@ class Viewport : public Node {
 		SCALING_3D_MODE_BILINEAR,
 		SCALING_3D_MODE_FSR,
 		SCALING_3D_MODE_FSR2,
+		SCALING_3D_MODE_SGSR1,
```
Adding an enum value in the middle means that all following enumerators change value. This is a breaking change that can easily be addressed by appending new constants at the end.

HH45137 replied:
Fixed. Thank you for the reminder.

Bromeon reviewed (dismissed):
I can't comment on the feature itself, but on a technical note, this contains some unnecessary breaking changes -- see also warnings in `4.4-stable.expected` file.

HH45137 committed:
Fix the code formatting

HH45137 commented on misc/extension_api_validation/4.4-stable.expected:
```diff
@@ -321,3 +321,14 @@ Validate extension JSON: Error: Field 'classes/RichTextLabel/methods/add_image/a
+Validate extension JSON: Error: Field 'classes/RenderingServer/enums/ViewportScaling3DMode/values/VIEWPORT_SCALING_3D_MODE_METALFX_SPATIAL': value changed value in new API, from 3.0 to 4.
+Validate extension JSON: Error: Field 'classes/RenderingServer/enums/ViewportScaling3DMode/values/VIEWPORT_SCALING_3D_MODE_METALFX_TEMPORAL': value changed value in new API, from 4.0 to 5.
+Validate extension JSON: Error: Field 'classes/Viewport/enums/Scaling3DMode/values/SCALING_3D_MODE_METALFX_SPATIAL': value changed value in new API, from 3.0 to 4.
+Validate extension JSON: Error: Field 'classes/Viewport/enums/Scaling3DMode/values/SCALING_3D_MODE_METALFX_TEMPORAL': value changed value in new API, from 4.0 to 5.
```
The issue has been fixed.

HH45137 committed:
Fix the compatibility issues of the renderer's compatibility mode

1. Shader compatibility issue

2. Crashes when selecting the MetalFX upsampling method in the compatibility mode

HH45137 commented:
I'm adapting to the mobile mode.

HH45137 commented:
@Bromeon Is it possible to merge the codes at present? :)

HH45137 commented:
> > @Bromeon Is it possible to merge the codes at present? :)
> 
> I already mentioned I can't comment on the feature itself 🙂 my review was only regarding compatibility breakage.
> 
> The PR will need review from the domain maintainers (rendering). As this is a new feature and Godot 4.5 enters a phase of focusing on bugfixing and regressions, this may need to wait a bit.


OK. Thank you for your help.

llama-nl commented:
> The PR will need review from the domain maintainers (rendering). As this is a new feature and Godot 4.5 enters a phase of focusing on bugfixing and regressions, this may need to wait a bit.

@HH45137 I think @clayjohn can review this.
